src/app/campaign/store/actions/campaign-action.service.ts

```diff
@@ -15,12 +16,54 @@ import {
 import { Moment } from 'moment';
 
 @Injectable()
-export class CampaignActionService {
-  constructor(private store: Store<any>) {}
+export class CampaignActionService implements OnDestroy {
+  // use a Subject to filter form updates through a pipe to dispatch form changes and load preview
+  flightFormValueChanges = new Subject<{ flight: Flight; changed: boolean; valid: boolean }>();
+  flightFormValueChangesSubscription: Subscription;
+  flightFormValueChangesStream = this.flightFormValueChanges.pipe(
+    // map form field values to a proper Flight
+    map(({ flight, changed, valid }) => {
+      return {
+        flight: {
+          ...flight,
+          uncapped: flight.uncapped || false,
+          totalGoal: flight.uncapped ? 0 : +flight.totalGoal,
+          dailyMinimum: flight.uncapped ? 0 : +flight.dailyMinimum
+        },
+        changed,
+        valid
+      };
+    }),
+    // if preview params changed, dispatch loadFlightPreview
+    withLatestFrom(this.store.pipe(select(selectRoutedFlight))),
+    tap(([formState, flightState]) => {
+      if (
+        this.hasPreviewParams(formState.flight) &&
+        this.isDateRangeValid(formState.flight) &&
+        this.havePreviewParamsChanged(flightState.localFlight, formState.flight)
+      ) {
+        this.loadFlightPreview(formState.flight);
+      }
+    }),
+    // guard flight form updates
+    filter(([formState, flightState]) => {
+      // are we still dealing with these form updates that aren't actual updates?? maybe just on route change at this point
+      const nameChanged = flightState.localFlight.name !== formState.flight.name;
+      const previewParamsChanged = this.havePreviewParamsChanged(flightState.localFlight, formState.flight);
+      if (!nameChanged && !previewParamsChanged) {
+        console.warn('unchanged?', { formState }, { flightState });
```

Comment:
I do get this warning everytime I change the TotalGoal or DailyMin fields using the up/down arrows.  Or hitting backspace a bunch of times.

Reply:
k, seems proof we still need the guard for whatever reason. That stinks because every new form field we add will have to be added to that guard, and they are just getting more complex from here out. 

I think I should remove or comment out this warning.

It does seem like `previewParamsChanged` would be `true` when you use the up/down arrows or backspace because it checks all the form params for changes that aren't name.

Frustrating.

Reply:
Seems like key events emit twice, for some reason?  So the 2nd time, preview params _haven't_ changed (and we're in-progress loading that preview).

Reply:
That makes me uncomfortable, and I don't like it. It's better and things are going in the right direction, but I suspected this might not solve all of our form problems.

Reply:
+1.  mo forms, mo problems.